include/aspect/melt.h

```diff
@@ -0,0 +1,329 @@
+        }
+
+        /**
+         * Compaction viscosity values $\xi$ at the given positions.
+         * This parameter describes the resistance of the solid matrix
+         * in a two-phase simulation to dilation and compaction.
+         */
+        std::vector<double> compaction_viscosities;
+
+        /**
+         * Fluid (melt) viscosity values $\eta_f$ at the given positions.
+         */
+        std::vector<double> fluid_viscosities;
+
+        /**
+         * Permeability values $k$ at the given positions.
+         */
+        std::vector<double> permeabilities;
+
+        /**
+         * Fluid (melt) density values $\rho_f$ at the given positions.
+         */
+        std::vector<double> fluid_densities;
+
+        /**
+         * An approximation for the fluid (melt) density gradients
+         * $\nabla \rho_f$ at the given positions. These values are
+         * required for compressible models to describe volume changes
+         * of melt in dependence of pressure, temperature etc.
+         */
+        std::vector<Tensor<1,dim> > fluid_density_gradients;
+
+        void average (const MaterialAveraging::AveragingOperation operation,
+                      const FullMatrix<double>  &projection_matrix,
+                      const FullMatrix<double>  &expansion_matrix)
+        {
+          average_property (operation, projection_matrix, expansion_matrix,
+                            compaction_viscosities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            fluid_viscosities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            permeabilities);
+          average_property (operation, projection_matrix, expansion_matrix,
+                            fluid_densities);
+
+          // The fluid density gradients are unfortunately stored in reverse
+          // indexing and averaging is not implemented for tensors (only for
+          // doubles). It's also not quite clear whether these should
+          // really be averaged, so avoid this for now
+        }
+    };
+
+    /**
+     * Base model for material models that implement a melt fraction function.
+     * This is used to compute some statistics about the melt fraction.
+     */
+    template <int dim>
+    class MeltFractionModel
+    {
+      public:
+        virtual void melt_fractions (const MaterialModel::MaterialModelInputs<dim> &in,
+                                     std::vector<double> &melt_fractions) const = 0;
+    };
+
+  }
+
+  /**
+   * Creates additional material model output objects that are
+   * needed for a simulation, and attaches a pointer to them to the
+   * corresponding vector in the MaterialModel::MaterialModelOutputs
+   * structure.
+   */
+  template <int dim>
+  void create_melt_material_outputs(MaterialModel::MaterialModelOutputs<dim> &output);
+
+
+  namespace Assemblers
```

Comment:
It's not wrong to put these function declarations into the header file, but it is probably also not necessary because you're only going to reference them from a single `.cc` file if I anticipate correctly what comes further down in the patch. You may be able to reduce the size of the file somewhat by putting all of this into the `.cc` file. (Or you may decide that you like it better here, because, for example, it shows up in the doxygen documentation this way.)